# Built-in functions and containers

## Course: Programming and Data Management (EDI 3400)

### *Vegard H. Larsen (Department of Data Science and Analytics)*

- - We will also learn how to use the `help()` function to get information about functions1. Built in functions

# 1. Built in functions

Built-in functions are pre-defined operations that come with the language, offering a plethora of utilities without the need for external libraries or additional code. They simplify common tasks, making programming more efficient and intuitive. For instance, `print()` displays information to the console, `len()` returns the length of an item, and `int()` or `str()` convert values between numerical and textual data types. Using these functions, a beginner can accomplish significant tasks without reinventing the wheel. Python's rich assortment of built-in functions, ranging from mathematical operations to data type manipulations, equips students with powerful tools that streamline code, enhancing readability and efficiency. 

- A built-in function is a function that is available without importing any additional libraries.
- Some useful built-in functions are: `print()`, `input()`, `type()`, `len()`, `round()` and `dir()`
- In lecture 5 we will learn how we can make our own functions

## Examples of usage of built in functions:

By now we have seen both `print()` and `type()` many times. We have also used `int()` for type conversion. 

Let's look at some other functions:

In [110]:
# Getting help
len?

In [113]:
# Compute the length of a list
len({1, 2, 3, 3})

3

In [114]:
round?

In [116]:
# Round of a number
round(3.1425, 2)

3.14

## The `input()` function

In [117]:
input() # This function let the user give an input from the keyboard

test


'test'

In [118]:
input('Hi user, enter a string here: ') # We can give the user some instructions

Hi user, enter a string here: test


'test'

In [119]:
user_input = input('Hi user, enter a number here: ') # Assign the input to a variable

Hi user, enter a number here: vegard


In [120]:
user_input  # The input is always a string-type

'vegard'

## Some other built-in functions:

In [9]:
# The absolute value of a number
abs(-10)

10

In [123]:
# Sum the element of an iterable object
sum([1,2])

3

In [11]:
# Type cast to a float
float(34)

34.0

In [125]:
# Get help
#help(int)

# 2. Objects have methods

Objects are instances of data types or classes, and these objects often come equipped with built-in functions called "methods" that perform actions or tasks specific to that object. Think of objects like smartphones: just as different smartphone models have unique apps or features, different objects in Python have specific methods. For example, a string object (e.g., `"hello"`) has methods like `.upper()` to convert it to uppercase, resulting in `"HELLO"`. Similarly, a list object (e.g., `[1, 2, 3]`) has methods such as `.append(4)` to add the number `4` to its end. These methods are invoked using the dot (`.`) notation.

## Methods vs. Functions

#### Methods 
- Methods are associated with the objects of the class they belong to.	
- A method is called 'on' an object. We cannot invoke it just by its name

#### Functions 
- Functions are not associated with any object.
- We can invoke a function just by its name.

## Almost everything in Python is an object

- An object can have its own methods (functions)
- - We can access these methods by using dot-notation
    - `object_name.{method}`
- We can see all the methods that belong to an object with the built-in `dir()` function

## What are the methods for an object of type `str`

In [109]:
print(dir(str))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


In [126]:
str.capitalize?

In [129]:
a_string = 'this is a string'
a_string.capitalize?

## Tab Completion

- Many text editors as well as the Jupyter Notebook can automatically fill in partially typed code

In [131]:
'this is a string, try using tab completion'.capitalize()

'This is a string, try using tab completion'

In [18]:
a_string = 'Apple'
a_number = 42.0

In [132]:
a_number.is_integer()

True

# 3. Lists

Lists are versatile, ordered collections that can hold a mixture of data types, making them invaluable for organizing and managing data. Represented by items enclosed in square brackets (e.g., `[1, 'apple', 3.14]`), lists are mutable, meaning their contents can be modified after creation. You can append items using the `.append()` method, remove them with the `.remove()` method, or access specific items by their index, with indexing starting from zero. For instance, in the list `fruits = ['apple', 'banana', 'cherry']`, `fruits[0]` would retrieve 'apple'. Understanding lists is foundational as they are often employed in loops, functions, and other core programming constructs, serving as dynamic storage structures that adapt to varying data needs.

## The built in lists-type

- One of the most versatile data types that allow us to work with multiple elements at once.
- A list can be created by using square brackets: `a_new_list = []`
- A list is mutable meaning you can modify a list after it is created

In [134]:
# A list can consist of different types of elements
example_list = ['this_is_a_string', 45, 3.121212, ['a', 'b']]

In [135]:
# Can also use the list built in method (requires an iterable as input)
list('abc')

['a', 'b', 'c']

## Accessing, assigning and changing elements in a list

- We can access the elements of the list by passing an integer to the list:

In [136]:
example_list[0]  # This is how we access the first element. Python counts from zero.

'this_is_a_string'

In [138]:
example_list[3][1]  # This is how we access the fourth element 

'b'

- We can also assign values to elements of the list

In [139]:
example_list[0] = 4 
example_list

[4, 45, 3.121212, ['a', 'b']]

## The length of a list

In [25]:
len(example_list)

4

In [141]:
example_list[3]

['a', 'b']

## Sorting lists

In [27]:
# Sorting numbers
a_list = [4, 6.9, 7, 1.5, 9]
a_list.sort() # Sorting in place
print(a_list)

[1.5, 4, 6.9, 7, 9]


In [142]:
# Sorting strings
b_list = ['this', 'is', 'a', 'list', 'of', 'words']
b_list.sort() # Sorting by length of word
print(b_list)

['a', 'is', 'list', 'of', 'this', 'words']


## List indexing and slicing

- We access the `i`´th element in the list as `list_name[i]` 
- We can also access parts of the list by selecting a `start` and `stop` index as `list_name[start:stop]`
- Note that here the `stop` index is not included when slicing

In [144]:
my_list = [0, 1, 2, 3, 4]
my_list[1:5]

[1, 2, 3, 4]

## List indexing and slicing, cont.

- How can we include the last element? 

In [145]:
my_list[:]  # If we drop the stop index we include the last element

[0, 1, 2, 3, 4]

- If we don't know the length of the list we can access the last element by counting from the back

In [148]:
my_list = [0, 1, 2, 3, 4]
my_list[-3:]              # This notation is giving us the last element of the list

[2, 3, 4]

# List indexing and slicing, cont.

We can also set a step value so we jump through the list by this step

`a[start:stop:step]` - start through not past stop, by step

In [32]:
a_new_list = list('This is a string that will be converted to a list!')
print(a_new_list)

['T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', ' ', 's', 't', 'r', 'i', 'n', 'g', ' ', 't', 'h', 'a', 't', ' ', 'w', 'i', 'l', 'l', ' ', 'b', 'e', ' ', 'c', 'o', 'n', 'v', 'e', 'r', 't', 'e', 'd', ' ', 't', 'o', ' ', 'a', ' ', 'l', 'i', 's', 't', '!']


In [149]:
a_new_list[1:30:5]

['h', 's', 't', ' ', ' ', ' ']

## Methods for Lists

If `a` is a list, there exists methods that can be performed on `a`. Some examples:

| Method | Description |
| ----------- | ----------- |
| `a.append(arg)` | Append object to the end of the list  |
| `a.remove(arg)` | Remove first occurrence of value      |
| `a.count(arg)`  | Return number of occurrences of value |
| `a.clear()`     | Remove all items from list            |

## Using the `list` methods

In [34]:
city_list = ['Oslo'] # Create a list with one element

In [35]:
city_list.append('Bergen') # Add an element to the list 

In [36]:
city_list

['Oslo', 'Bergen']

In [37]:
city_list.remove('Oslo') # Remove an element from list

In [38]:
city_list

['Bergen']

In [39]:
print(city_list.count('Bergen')) # Count a specific element ('Bergen') in the list

1


## Assigning lists with `=` does not make a copy
- The variable name is only pointing to the actual list object

In [154]:
first_list = ['a', 'b', 'c']

In [155]:
second_list = first_list.copy()    # can make a copy with first_list.copy()

In [156]:
second_list.clear()

In [157]:
print(first_list)

['a', 'b', 'c']


# 4. Tuples

Tuples are akin to lists in that they are ordered collections of items; however, there are crucial differences. Enclosed in parentheses (e.g., `(1, 'apple', 3.14)`), tuples are immutable, meaning once they're created, their content cannot be altered—no additions, deletions, or modifications. This fixed nature makes tuples ideal for representing things that shouldn't change, such as days of the week or coordinates on a map. Accessing items in a tuple is just like with lists, using indexing; for example, in the tuple `colors = ('red', 'green', 'blue')`, `colors[1]` would yield 'green'.

## The built-in tuple-type

- Tuples are similar to lists, allowing us to collect values in an ordered way
- Tuples differ from lists in that they are unchangeable (immutable),
    * We cannot change, add or remove items after the tuple has been created

## Creating tuples

- Tuples can be created by using round brackets

In [45]:
tuple1 = (1, 5, 'text', [1,2,3])   # A tuple created with round brackets

- Can skip the round brackets, but the code might be harder to read

In [158]:
tuple2 = 3, 5, 5, 9, 11, 45, 67    # A tuple created without round  brackets

In [47]:
# We can access the elements in the tuples as be do with lists
print(tuple1[3])     # We can select elements
print(tuple2[5:])    # We can make a slice of the tuple  

[1, 2, 3]
(45, 67)


## What happens if we try to assign a value to an existing `tuple`?

In [159]:
tuple1[2] = 5  # We are not allowed to assign values to a tuple

TypeError: 'tuple' object does not support item assignment

## Methods for  tuples 

- Since a `tuple` cannot be changed after it is created, there are very few built-in methods for a  `tuple`
- If `t` is a `tuple`, these are some methods that can be performed on `t`

| Method | Description |
| ----------- | ----------- |
| `t.count(value)`  | Return number of occurrences of value |
| `t.index(value)`  | Return first index of value|

## Example of using the tuple-methods

In [49]:
tuple2 = (3, 5, 5, 9, 11, 45, 67)

In [50]:
tuple2.count(5) # Count the number of occurrences of the number 5 in tuple2

2

In [160]:
tuple2.index(5) # What is the location of the number 45 in tuple2

1

## Unpacking tuples

In [193]:
# Store some data about some cities using tuples 
oslo = ('Norway', 630_000, True)
bergen = ('Norway', 270_000, False)
new_york = ('US', 8_380_000, False)

In [194]:
country_oslo, population_oslo, is_capital_oslo = oslo

In [195]:
print(population_oslo)

630000


In [196]:
country_bergen, population_bergen, is_capital_bergen = bergen
country_new_york, population_new_york, is_capital_new_york = new_york

In [197]:
print(population_new_york)

8380000


# Unpacking many elements

In [207]:
oslo = ('Norway', 630_000, True, 'O', 's', 'l', 'o')

In [210]:
6.3e5 == 630_000

True

In [202]:
country_oslo, population_oslo, is_capital_oslo, *other = oslo

In [203]:
other

['O', 's', 'l', 'o']

In [204]:
country, *the_middle, last_letter = oslo

In [205]:
the_middle

[630000, True, 'O', 's', 'l']

In [206]:
last_letter

'o'

# 5. Strings

Strings represent sequences of characters and are vital for handling text-based data. Encased within single (`'...'`), double (`"..."`), or even triple (`'''...'''` or `"""..."""` for multi-line strings) quotes, they are more than simple text labels. Strings possess a wealth of built-in methods, like `.upper()` to capitalize characters or `.replace()` to substitute portions of the text. Just like lists and tuples, strings are ordered, allowing access to individual characters using indices: in the string `word = "python"`, `word[0]` refers to the character 'p'. However, unlike lists, strings are immutable, meaning their content cannot be directly altered once created. Grasping strings is essential, as text processing is a cornerstone in many programming endeavors, from user input to data analysis.

## Defining strings

* Strings can be created by using either single or double quotes

In [212]:
text1 = 'This is a string'
text2 = "This is another string"
text3 = "A 'string' with a quote"
text4 = 'Another "string" with a quote'

In [64]:
print(text3)

A 'string' with a quote


In [65]:
print(text4)

Another "string" with a quote


## Methods for  Strings

If `s` is a string, these are some methods that can be performed on `s`:

| Method | Description |
| ----------- | ----------- |
| `s.lower()`               | Returns the lowercase version of `s` |
| `s.upper()`               | Returns the uppercase version of `s` |
| `s.capitalize()`          | Return a capitalized version of `s`  |
| `s.replace('old', 'new')` | Return a copy with all occurrences of the sub-string old replaced by new |
| `s.split()`               | Return a list of the words in the string |

## Using the `string` methods

In [219]:
test_string = 'this Is a STRING.'

In [220]:
test_string.lower()  # Turn all the letters in the string into lowercase 

'this is a string.'

In [221]:
test_string.capitalize() # Capitalizes the first letter and turn the rest into lowercase

'This is a string.'

In [222]:
test_string

'this Is a STRING.'

In [218]:
test_string.replace('STRING', 'text') # Replace the 'STRING' with 'text'

'this Is a text.'

In [70]:
test_string.split() # Turn the string into a list

['this', 'Is', 'a', 'STRING.']

## Special characters in strings

- \n - Newline

- \t - Horizontal tab

- \\' - Single Quote

- \\" - double quote

- \\\ - Backslash

- \v - vertical tab


In [223]:
print('What\'s your name?') 

What's your name?


In [224]:
# Alternatively
print("What's your name?") 

What's your name?


In [225]:
print('1\t2\t3\n4\t5\t6\n7\t8\t9')

1	2	3
4	5	6
7	8	9


## f-strings

- f-strings are strings that have an f at the beginning and curly braces containing any valid Python expressions that will be replaced with their values.

In [74]:
salary = 30000 # Current salary
increase = 0.04 # Increase for next year is 4%

In [226]:
# Printing next years salary adding regular strings together 
print('The salary will increase by 4.0% to ' + str(salary*(1+increase)) + ' next year.')

The salary will increase by 4.0% to 31200.0 next year.


In [231]:
# Using the f-string
print(f'The salary will increase by {increase:.1%} to {salary*(1+increase)} next year.')

The salary will increase by 4.0% to 31200.0 next year.


# 6. Dictionaries 

Dictionaries are powerful data structures that store pairs of keys and values, facilitating quick data retrieval based on a unique key. Unlike lists and tuples, which are ordered by index, dictionaries use curly braces `{...}` and are inherently unordered. For instance, in the dictionary `student = {'name': 'John', 'age': 18, 'grade': 'A'}`, 'name', 'age', and 'grade' are keys, each associated with specific values. To access the value for 'name', one would use `student['name']`, obtaining the result 'John'. Since dictionaries are mutable, their content can be modified post-creation, enabling dynamic data storage.

## Using dictionaries

- Unlike lists, tuples and strings, which are indexed by a range of numbers, dictionaries are indexed by keys
- A key can be any immutable type
- A dictionary is as a set of key: value pairs, with the requirement that the keys within one dictionary are unique
- Dictionaries can be created by inserting a comma-separated list of key:value pairs inside curly brackets `{}` 
- Fast way of getting the value when you know the key

## Creating a dictionary

In [233]:
oslo = {'country': 'Norway', 'population': 630_000, 'is_capital': True}

In [234]:
print(oslo)

{'country': 'Norway', 'population': 630000, 'is_capital': True}


In [235]:
# Can access the elements by the key
oslo['population']

630000

In [236]:
# Can check if key is in dict
'country' in oslo

True

## Methods for Dictionaries

If `d` is a dictionary, these are some methods that can be performed on `d`:

- `d.keys()`
- `d.values()`
- `d.clear()`

## Using the dictionary methods

In [238]:
oslo.keys()

dict_keys(['country', 'population', 'is_capital'])

In [239]:
oslo.values()

dict_values(['Norway', 630000, True])

In [240]:
oslo.clear()

In [241]:
oslo

{}

In [85]:
oslo.

# 7. Sets

Sets are unique collections that store unordered items, similar in concept to mathematical sets. Defined using curly braces `{...}`, like dictionaries, or via the built-in `set()` function, sets inherently eliminate duplicate values. For example, the set `{1, 2, 3, 3, 4}` would automatically discard the extra `3`, resulting in `{1, 2, 3, 4}`. Sets also offer powerful operations like union (`|`), intersection (`&`), and difference (`-`), mirroring classical set theory. A notable characteristic of sets is their inability to store mutable types, such as lists, but they can handle immutable types like tuples and strings. 

## Sets

- A set is an unordered collection with no duplicate elements.
- Sets are unordered.
- Set items are unchangeable, but you can remove items and add new items to the set.

## Defining sets

- A set can be created in two ways: 
    1. Can be defined with the built-in `set()` function
    2. Can be defined with curly braces `{}`:

In [244]:
first_set = set(['apple'])
print(first_set)

{'apple'}


In [87]:
second_set = {'apple', 'orange', 'banana', 'apple', 'banana'}
print(second_set)

{'orange', 'banana', 'apple'}


## Set operations

In [88]:
A = {1, 2, 3, 4, 5, 6}

B = {5, 6, 7, 8, 9, 10}

### Union
The set of all elements in the whole collection (both A and B).

$A \cup B =  \{1, 2, 3, 4, 5, 6, 7, 8, 9, 10\}$

### Intersection 
The set containing all elements of A that also belong to B

$A \cap B = \{5, 6\}$

In [242]:
A.union(B)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

In [90]:
A.intersection(B)

{5, 6}